lib/ueberauth/strategy.ex

```diff
@@ -318,4 +336,96 @@ defmodule Ueberauth.Strategy do
     auth = apply(strategy, :auth, [conn])
     Plug.Conn.assign(conn, :ueberauth_auth, auth)
   end
+
+  defp maybe_run_handle_callback(true, conn, strategy) do
+    run_handle_callback(conn, strategy)
+  end
+
+  defp maybe_run_handle_callback(_ignore_state_param, conn, strategy) do
+    if state_param_matches?(conn) do
+      run_handle_callback(conn, strategy)
+    else
+      add_state_mismatch_error(conn, strategy)
+    end
+  end
+
+  defp state_param_matches?(conn) do
+    state_cookie = get_state_cookie(conn)
+    param_cookie = conn.params["state"]
+
+    case {state_cookie, param_cookie} do
+      {nil, nil} -> false
+      {value, value} -> true
+      {_, _} -> false
+    end
+  end
+
+  defp add_state_mismatch_error(conn, strategy) do
+    conn
+    |> Helpers.set_errors!([
+      %Error{message_key: :csrf_attack, message: "Cross-Site Request Forgery attack"}
+    ])
+    |> run_handle_cleanup(strategy)
+  end
+
+  defp run_handle_request(conn, strategy) do
+    apply(strategy, :handle_request!, [conn])
+  end
+
+  defp run_handle_callback(conn, strategy) do
+    strategy
+    |> apply(:handle_callback!, [conn])
+    |> handle_callback_result(strategy)
+    |> run_handle_cleanup(strategy)
+  end
+
+  defp run_handle_cleanup(conn, strategy) do
+    # TODO: should we remove the cookie always?
+    conn = remove_state_cookie(conn)
+    apply(strategy, :handle_cleanup!, [conn])
+  end
+
+  defp maybe_add_state_param(conn, strategy) do
+    strategy
+    |> get_ignores_csrf_attack_option()
+    |> add_state_param(conn)
+  end
+
+  defp get_ignores_csrf_attack_option(strategy) do
+    strategy
+    |> apply(:default_options, [])
+    # TODO: should we force them to turn it off
```

Comment:
I think we should default to the safest option. We can release all new versions of our strategies to coincide with this release if need be.